engines/sci/graphics/gfxdrivers.cpp

```diff
@@ -0,0 +1,508 @@
+
+namespace Sci {
+
+Common::Point GfxDriver::getMousePos() const {
+	return g_system->getEventManager()->getMousePos();
+}
+
+GfxDefaultDriver::GfxDefaultDriver(uint16 screenWidth, uint16 screenHeight) : GfxDriver(screenWidth, screenHeight, 0, 1) {
+	switch (g_sci->getResMan()->getViewType()) {
+	case kViewEga:
+		_numColors = 16;	// QFG PC-98 with 8 colors also reports 16 here
+		break;
+	case kViewAmiga:
+		_numColors = 32;
+		break;
+	case kViewAmiga64:
+		_numColors = 64;
+		break;
+	case kViewVga:
+	case kViewVga11:
+		_numColors = 256;
+		break;
+	default:
+		break;
+	}
+
+	if (_numColors == 0)
+		error("GfxDefaultDriver: Unknown view type");
+}
+
+void GfxDefaultDriver::setPalette(const byte *colors, uint start, uint num) {
+	g_system->getPaletteManager()->setPalette(colors, start, num);
+}
+
+void GfxDefaultDriver::copyRectToScreen(const byte *src, int pitch, int x, int y, int w, int h) {
+	g_system->copyRectToScreen(src, pitch, x, y, w, h);
+}
+
+void GfxDefaultDriver::replaceCursor(const void *cursor, uint w, uint h, int hotspotX, int hotspotY, uint32 keycolor) {
+	CursorMan.replaceCursor(cursor, w, h, hotspotX, hotspotY, keycolor);
+}
+
+SCI0_DOSPreVGADriver::SCI0_DOSPreVGADriver(int numColors, int screenW, int screenH, int horizontalAlignment) : GfxDriver(screenW, screenH, numColors, horizontalAlignment), _palNeedUpdate(true), _colors(nullptr), _compositeBuffer(nullptr) {
+	_compositeBuffer = new byte[screenW * screenH]();
+}
+
+SCI0_DOSPreVGADriver::~SCI0_DOSPreVGADriver() {
+	delete[] _compositeBuffer;
+}
+
+bool SCI0_DOSPreVGADriver::checkDriver(const char *const *driverNames, int listSize) {
+	Common::String missing;
+	while (listSize-- && *driverNames) {
+		if (Common::File::exists(*driverNames))
+			return true;
+		if (!missing.empty())
+			missing += " or ";
+		missing += "'" + Common::String(*driverNames) + "'";
+		++driverNames;
+	}
+	warning("Driver file %s not found. Starting game in EGA mode", missing.c_str());
+	return false;
+}
+
+void SCI0_DOSPreVGADriver::assignPalette(const byte *colors) {
+	_colors = colors;
+}
+
+void SCI0_DOSPreVGADriver::setPalette(const byte*, uint, uint) {
+	if (!_palNeedUpdate || !_colors)
+		return;
+	_palNeedUpdate = false;
+	byte *tmp = new byte[768]();
+	memcpy(tmp, _colors, _numColors * 3);
+	g_system->getPaletteManager()->setPalette(tmp, 0, 256);
+	delete[]tmp;
+}
+
+SCI0_EGADriver::SCI0_EGADriver() : SCI0_DOSPreVGADriver(16, 320, 200, 1) {
+	static const uint8 egaColors[] = {
+		0x00, 0x00, 0x00, 0x00, 0x00, 0xAA, 0x00, 0xAA, 0x00, 0x00, 0xAA, 0xAA,
+		0xAA, 0x00, 0x00, 0xAA, 0x00, 0xAA, 0xAA, 0x55, 0x00, 0xAA, 0xAA, 0xAA,
+		0x55, 0x55, 0x55, 0x55, 0x55, 0xFF, 0x55, 0xFF, 0x55, 0x55, 0xFF, 0xFF,
+		0xFF, 0x55, 0x55, 0xFF, 0x55, 0xFF, 0xFF, 0xFF, 0x55, 0xFF, 0xFF, 0xFF
+	};
+	assignPalette(egaColors);
+}
+
+void SCI0_EGADriver::copyRectToScreen(const byte *src, int pitch, int x, int y, int w, int h) {
+	byte *dst = _compositeBuffer;
+	pitch -= w;
+	for (int i = 0; i < h; ++i) {
```

Comment:
Is there a reason for having an intermediate buffer in EGA mode?

Reply:
See comment above. I made that at the very beginning and then just didn't modify it any more. I'll try remove the class and set the SCI0 EGA games to the default driver.